packages/batch-tx-searcher/scripts/utils.ts

```diff
@@ -0,0 +1,147 @@
+import { NodeInterface__factory } from "@arbitrum/sdk/dist/lib/abi/factories/NodeInterface__factory"
+import { Bridge__factory } from "@arbitrum/sdk/dist/lib/abi/factories/Bridge__factory"
+import { NODE_INTERFACE_ADDRESS } from "@arbitrum/sdk/dist/lib/dataEntities/constants"
+import { getL2Network, L2Network } from "@arbitrum/sdk"
+import { providers, BigNumber } from "ethers";
+import { NodeInterface } from "@arbitrum/sdk/dist/lib/abi/NodeInterface";
+
+export interface BlockRange {
+    start: number,
+    end: number
+}
+
+// Use binary search to find one of the target block.
+const binarySearchMatchedBlock =async (
+    batchNum: BigNumber,
+    latestBlockNum: number,
+    nodeInterface: NodeInterface,
+    l2network: L2Network
+): Promise<number> => {
+    const nitroGenesisBlock = l2network.nitroGenesisBlock
+    let high = latestBlockNum;
+    let low = nitroGenesisBlock;
+
+    while (low < high) {
+        const mid = Math.round((low + high)/2);
+        const currentBatch = await nodeInterface.findBatchContainingBlock(mid)
+        //When we find one, we can return directly to reduce the rpc call load.
+        if(currentBatch.eq(batchNum)) {
+            return mid;
+        } else if (currentBatch.gt(batchNum)) {
+            high = mid;
+        } else {
+            low = mid + 1;
+        }
+    }
+    return high == nitroGenesisBlock ? nitroGenesisBlock : high - 1;
+}
+
+//Given one of the matched block, then find all matched blocks.
+const findBlockRange =async (
+    batchNum: BigNumber,
+    blockNum: number,
+    latestBlockNum: number,
+    nodeInterface:NodeInterface,
+    l2network: L2Network
+): Promise<BlockRange> => {
+    let start = l2network.nitroGenesisBlock;
+    let end = latestBlockNum;
+
+    /**
+     * First try to find all blocks whose block height lower than this blocknum.
```

Comment:
You can modify the binary search to find boundaries of duplicated values, when you have an exact match, you check if x[m+1]==x[m]+1, if not you set l=m instead of returning, you will get the right boundary this way, then you can repeat the same for the left boundary with little modification.